File: 3. Behavioral Design Patterns/Iterator/Python/IteratorTemplate.py
# Iterator Pattern – Python Template

from abc import ABC, abstractmethod

class Iterator(ABC):
    @abstractmethod
    def has_next(self):
        pass
    
    @abstractmethod
    def next(self):
        pass

class Aggregate(ABC):
    @abstractmethod
    def create_iterator(self):
        pass

class ConcreteIterator(Iterator):
    def __init__(self, items):
        self.items = items
        self.index = 0
    
    def has_next(self):
        return self.index < len(self.items)
    
    def next(self):
        if self.has_next():
            item = self.items[self.index]
            self.index += 1
            return item
        return None

class ConcreteAggregate(Aggregate):
    def __init__(self):
        self.items = []
    
    def add(self, item):
        self.items.append(item)
    
    def create_iterator(self):
        return ConcreteIterator(self.items)

def main():
    aggregate = ConcreteAggregate()
    aggregate.add("Item1")
    aggregate.add("Item2")
    
    iterator = aggregate.create_iterator()
    while iterator.has_next():
        item = iterator.next()
        print(item)

if __name__ == "__main__":
    main()

File: 3. Behavioral Design Patterns/Iterator/Go/IteratorTemplate.go
package main

// Iterator Pattern – Go Template

type Iterator interface {
	HasNext() bool
	Next() interface{}
}

type Aggregate interface {
	CreateIterator() Iterator
}

type ConcreteIterator struct {
	items []interface{}
	index int
}

func NewConcreteIterator(items []interface{}) *ConcreteIterator {
	return &ConcreteIterator{items: items, index: 0}
}

func (ci *ConcreteIterator) HasNext() bool {
	return ci.index < len(ci.items)
}

func (ci *ConcreteIterator) Next() interface{} {
	if ci.HasNext() {
		item := ci.items[ci.index]
		ci.index++
		return item
	}
	return nil
}

type ConcreteAggregate struct {
	items []interface{}
}

func NewConcreteAggregate() *ConcreteAggregate {
	return &ConcreteAggregate{items: make([]interface{}, 0)}
}

func (ca *ConcreteAggregate) Add(item interface{}) {
	ca.items = append(ca.items, item)
}

func (ca *ConcreteAggregate) CreateIterator() Iterator {
	return NewConcreteIterator(ca.items)
}

func main() {
	aggregate := NewConcreteAggregate()
	aggregate.Add("Item1")
	aggregate.Add("Item2")
	iterator := aggregate.CreateIterator()
	for iterator.HasNext() {
		item := iterator.Next()
		_ = item
	}
}
